scan

```diff
@@ -91,74 +129,357 @@ def run(options=None):
         os.remove(result)
 
+    # Trigger the Lambda pipeline.
+    use_lambda = options.get("lambda", False)
+
+    # Store local errors/timing info, and if using Lambda, trigger the
+    # Lambda post-processing pipeline to get Lambda timing/usage info.
+    meta = options.get("meta", False)
+
     # Run through each scanner and open a file and CSV for each.
     handles = {}
     for scanner in scanners:
         name = scanner.__name__.split(".")[-1]  # e.g. 'pshtt'
         scanner_filename = "%s/%s.csv" % (utils.results_dir(), name)
         scanner_file = open(scanner_filename, 'w', newline='')
         scanner_writer = csv.writer(scanner_file)
-        scanner_writer.writerow(["Domain", "Base Domain"] + scanner.headers)
+
+        # Write the header row, factoring in Lambda detail if needed.
+        headers = prefix_headers + scanner.headers
+
+        # Local scan timing/errors.
+        if meta:
+            headers += local_headers
+
+        # Lambda scan timing/errors. (At this step, only partial fields.)
+        if meta and use_lambda:
+            headers += lambda_headers
+
+        scanner_writer.writerow(headers)
 
         handles[scanner] = {
+            'name': name,
             'file': scanner_file,
             'filename': scanner_filename,
-            'writer': scanner_writer
+            'writer': scanner_writer,
+            'headers': headers
         }
 
-    # The task wrapper that's parallelized using executor.map.
-    def process_scan(params):
-        scanner, domain, options = params
-
-        # A scanner can return multiple rows.
-        try:
-            rows = list(scanner.scan(domain, options))
-        except:
-            logging.warn(utils.format_last_exception())
+    # Experimental!
+    if use_lambda:
+        scan_method = perform_lambda_scan
+    else:
+        scan_method = perform_local_scan
 
-        if rows:
-            for row in rows:
-                if row:
-                    handles[scanner]['writer'].writerow([domain, utils.base_domain_for(domain)] + row)
+    # Store scan UUID.
+    logging.debug("[%s] Scan UUID." % scan_uuid)
 
     # Run each scanner (unique process pool) over each domain.
     # User can force --serial, and scanners can override default of 10.
+    durations = {}
     for scanner in scanners:
 
         if options.get("serial"):
             workers = 1
         elif hasattr(scanner, "workers"):
             workers = scanner.workers
         else:
-            workers = int(options.get("workers", 10))
-
+            workers = int(options.get("workers", default_workers))
+
+        # Enforce a local worker maximum as a safety valve.
+        workers = min(workers, global_max_workers)
+
+        # Scan environment, passed to all scanners (local or cloud).
+        environment = {}
+        environment['scan_method'] = {True: 'lambda', False: 'local'}[use_lambda]
+        environment['scan_uuid'] = scan_uuid
```

Comment:
The `scan_method` and `scan_uuid` are sent into all scan methods, including Lambda executions. The `scan_method` should be referenced sparingly by a scanner, so as to keep the code agnostic, but for example this is used inside the `sslyze` scanner to default to synchronous execution in Lambda (no multiprocessing) and parallel execution locally (to avoid long-running memory leaks) without forcing the user to keep track of those things.